### DISCUSSION TITLE
Toggling vi normal and insert mode?

### DISCUSSION THREAD

**Discussion by schlich** [nushell/nushell#13509] [Q&A] [answered]
I've scoured the docs, issues, discussions, and even reedline's docs and have not seen any references on how to toggle vi's normal and insert modes.  For reference, I use Helix as my primary editor and rarely touch vi/vim/neovim, definitely not in an editor.  Whether this is super common knowledge or not I think it'd be helpful to briefly describe how the vi modes behave in reedline! 

nushell for life

**Comment by Lockszmith-GH** [comment, marked as answer]
# tl;dr:

### Hit <kbd>ESC</kbd> to exit the default _insert mode_.

## The long version

<details> <summary> Searching for <a href=https://www.google.com/search?hl=en&q=readline toggle vi mode>readline toggle vi mode </a> directed me to GNU's bash Readline documentation <sub><kbd>👆</kdb></sub> </summary>

ℹ️The highlight is irrelevant, it the way Google renders it, not mine.

![image](https://github.com/user-attachments/assets/fafbf7e2-2922-471f-94eb-3a4df8f88806)

</details>

Below is the relevant excerpt for that [page](https://www.gnu.org/software/bash/manual/html_node/Readline-vi-Mode.html):

> When you enter a line in vi mode, you are already placed in __’insertion’__ mode, as if you had typed an <kbd>i</kbd>. Pressing <kbd>ESC</kbd> switches you into __‘command’__ mode, where you can edit the text of the line with the standard vi movement keys, move to previous history lines with <kbd>k</kbd> and subsequent lines with <kbd>j</kbd>, and so forth. 

To test it, I searched the default `nu` configuration for `emacs` and found that the relevant configuration to switch on **vi mode** is:

```nushell
$env.config.edit_mode = vi
```

<details> <summary> <h2> More info </h2> <sub><kbd>👆</kdb></sub></summary>

* 📑 The line from the bash documentation is a verbatim copy from the GNU Readline User Interface documentation, specificaly the [vi mode section.](https://tiswww.case.edu/php/chet/readline/rluserman.html#Readline-vi-Mode)

* 📦 ~~`nushell` is using [`rustyline`](https://github.com/kkawakam/rustyline) as the readline implementation, the [README](https://github.com/kkawakam/rustyline#user-content-actions) contains all the keyboard mappings implemented~~.  
    Actually, while there are traces of `rustlyline` in the code, it seems the implemenation now is nushell's own [`reedline`](https://github.com/nushell/reedline) project, (now I understand the rabit hole you went through).

* 🧬 Helix is awesome, but it's _not vi_ - so not all mappings match exactly.

* 👩‍🎓 If you're interested in an interactive way of learning default vi keymapings, I would recommend going through an [online interactive tutorial](https://duckduckgo.com/?q=interactive+vim+tutorial&t=ffab&ia=web) or just run the vi editor and going through [the tutorial](https://www.dwarmstrong.org/neovim/#2-launch) which will provide a good refrence to get comfortable with the basic controls.


 </details>

**Reply by schlich** [reply]
It's considerably easy to find things when you know what you're searching for -- don't really appreciate the tone!  The quoted excerpt is exactly the information I was looking for, and I think it would have taken me considerably longer than 5 minutes to find it.  My point is that I think including this information would be a valuable addition to the documentation.

**Reply by fdncred** [reply]
Please add whatever you think is valuable via a PR to our docs repo.

**Reply by Lockszmith-GH** [reply]
Sorry, was way too late on my end, I appologize for the tone. I'll try and edit it a bit later.
All I literally did was searching the terms in your question. The details were at the top of the search results.

**Reply by Lockszmith-GH** [reply]
I've updated the text, and added some more context.

One thing I did notice, you refer to it as _reedline_ (twice), that's a specific name of a rust implementation. The common library name is **readline** - that might be what stumped your searches.

I agree, pointing to source documentation with a simple quote makes things easier for anyone trying to grasp "a lot of new".

Maybe `reedline` needs to take cue from `rustyline` on the documentation side.

Again, sorry for the tone. I was just tired and grumpy.
